14 commit offsets

martanit opened the pull request:
Changes:
- cpp:
  - Commit - exposing `rd_kafka_commit`: manually commit offsets. 
  - SetTopicPartitionList - exposing `rd_kafka_topic_partition_list_add`: has been modified to allow specifying also the partition id (previously, it was used only when subscribing the consumer to a topic list, for which the partition argument is ignored and therefore set to 0).
   - SetOffset - exposing `rd_kafka_topic_partition_list_set_offset`: provided a `rd_kafka_topic_partition_list_t`, a topic, a partition, and an offset, this function set the offset.

- APL:
  - kafka.commit
  - kafka.set_offset: return an error if the specified combination of topic and partition does not exists in the provided `rd_kafka_topic_partition_list_t`.

and in Consumer.aplc
  - set_topic_partition: has been modified to accept the partition id.
  - subscribe: has been modified to pass "0" as partition id.
  - set_offset
  - commit: at the moment commit **does not** take any argument: the topic partition list is 0 (which means "commit on all the subscribed topic/partition") and asynch is 0 (which translate to "commit synch").

To test manual commit offsets on a specific combination of topic and partition, modify the `commit` function in Consumer.aplc to 
```
    ∇ r←commit tplist
      :Access public
      r←kafka.commit cons tplist 0
    ∇
```
then:
`tp_to_commit←consumer.topic_partition`
`consumer.set_topic_partition(tp_to_commit, 'topic' p)` 
(can add other topics or partitions calling again `consumer.set_topic_partition`)
will create a new topic partition list with 'topic'(s) and partition "p"
`consumer.set_offset(tp_to_commit, 'topic' p offset)`
will set the offset "offset" for the specified 'topic' and partition "p".
`consumer.commit tp_to_commit`
will commit to that offest.

cybearinsight commented:
I think the changes looks good, maybe we could disable the option to call Commit with Async in the kafka.cpp until we have the onCommit function set. 
Or 
For the producer we set the onDelivery call back when we create the Producer, maybe we could do the same for the consumer so the onCommit is always set.

cybearinsight commented:
maybe we need to extent the consume result to also include the offset

martanit commented:
> I think the changes looks good, maybe we could disable the option to call Commit with Async in the kafka.cpp until we have the onCommit function set. Or For the producer we set the onDelivery call back when we create the Producer, maybe we could do the same for the consumer so the onCommit is always set.

Fixed in https://github.com/Dyalog/kafka/pull/15/commits/b5d313f3093d93d44aa18d6e3044c77a8360d051.

> maybe we need to extent the consume result to also include the offset

Added in https://github.com/Dyalog/kafka/pull/15/commits/ac5fc8f513a462f3166d56359083b211e95dbaa5.
